### ISSUE TITLE
prediction.shape different from input.shape

### ISSUE THREAD

**Issue by AlibekJ** [jakeret/tf_unet#41]
Here is my code
```
from tf_unet import unet, util, image_util
data_provider = image_util.ImageDataProvider("images3/*.png", data_suffix='.png', mask_suffix='_mask.png')
net = unet.Unet(layers=2, features_root=128, channels=3, n_class=2) # 
trainer = unet.Trainer(net)
path = trainer.train(data_provider, "train/", dropout=0.5, training_iters=32, epochs=1, display_step=16)
x_test = a._load_file("images4/0393_1.png")
x_test = a._process_data([x_test])
prediction = net.predict(path, x_test)
```
`prediction.shape` returns (1, 240, 240, 2)
`x_test.shape` returns (1, 256, 256, 3),
prediction image size is smaller

If layers=1, prediction shape is as 252x252. 
At layers=5 I get prediction of something like 96x96. 
And it is not scaled down version of the image, it just cuts the center of the image instead.

Is it a bug or a feature? How can I get a full shape while using more layers?

**Comment by jakeret** [comment]
This is due to the network architecture proposed by `Ronneberger et al. <https://arxiv.org/pdf/1505.04597.pdf>`_ and the resulting resolution loss. So, I guess it's a feature.. ;-)

I was still getting good results when mirroring the image on the borders

**Comment by ysyyork** [comment]
Hi @jakeret , would you mind sharing how you do border mirroring? The output is too small which makes my dataset almost useless and the result is really bad. I think I need to do border mirroring first. Thanks a lot!

**Comment by ysyyork** [comment]
Hi, @jakeret How much borders I should set for each image? Any formula to calculate the width of the border? I tried to add borders equal to the difference between the output and input but then it seems that it's cutting different for different image size. So the resulting image size is not exactly the same as the input

**Comment by AlibekJ** [comment]
@ysyyork 
I would suggest to feed images of exactly the same size. Depends on your application of course, but I would scale the input image to whatever unet was trained on and then scale back afterwards.

Here is how I add borders using OpenCV
```
img = cv2.copyMakeBorder(img, top=bordersize, bottom=bordersize, left=bordersize, right=bordersize,
                                     borderType=cv2.BORDER_CONSTANT, value=[0, 0, 0])
```

**Comment by Adopteruf** [comment]
I guess there is formula to calculate the increased width or height because the exact convolution/deconvolution procedures can be seen at least. 
But spending brain power on that is not always necessary if you just want to find one map, for example, what height x width maps to 224 x 200. 
You could just try any possible combinations by following your intuition, which is stupid but efficient, to some extent. Since the maps of height and widths are irrelevant due to the principle of convolution/deconvolution, it turns to be much more simple.
For example, with a few attempts, I obtain a map: (265, 240) --> (224, 200).
Besides, the method proposed by @AlibekJ is more standard. Good Job!!!
But in that method, you'd better set the input-size of the U-net large enough so that the output-size is larger than the size of your real input image which means you can add borders first and crop borders later. Unless, I think it will affect the segmentation on the border of your real image.
What's more, I think the functions: def crop_to_shape() [for directly cropping the real Y of the real X in order to fit the prediction]in util.py is not appropriate enough since the issues on segmentation on the border of the real image will also rise. @jakeret
